core/src/main/scala/pureconfig/DerivedWriters.scala

```diff
@@ -108,8 +119,23 @@ trait DerivedWriters1 {
     }
   }
 
+  // Auxiliary type to make sure we always return `ConfigList` when writing HLists.
+  private[pureconfig] trait HListConfigWriter[T <: HList] extends ConfigWriter[T] {
```

Comment:
I've removed `HListConfigWriter` in 6bb77972746e7f4b39cd771a4e795eaee053741e and now throw an exception whenever the `to` method doesn't return the appropriate type.